File: tests/core/contractors/codac2_tests_CtcCartProd.py
#!/usr/bin/env python

#  Codac tests
# ----------------------------------------------------------------------------
#  \date       2024
#  \author     Simon Rohou
#  \copyright  Copyright 2024 Codac Team
#  \license    GNU Lesser General Public License (LGPL)

import unittest
from codac import *

class TestCtcCartProd(unittest.TestCase):
  
  def tests_CtcCartProd(self):
    
    v = VectorVar(1)
    c1 = CtcInverse(AnalyticFunction([v], v), IntervalVector([[-1,1]]))
    c2 = CtcInverse(AnalyticFunction([v], v), IntervalVector([[5,6]]))

    a = IntervalVector(1)
    b = IntervalVector(1)
    c1.contract(a)
    self.assertTrue(a == IntervalVector([[-1,1]]))
    c2.contract(b)
    self.assertTrue(b == IntervalVector([[5,6]]))

    x = IntervalVector(2)
    c3 = CtcCartProd(c1,c2)
    c3.contract(x)
    self.assertTrue(x == IntervalVector([[-1,1],[5,6]]))

    y = IntervalVector(2)
    c4 = cart_prod(c1,c2)
    c4.contract(y)
    self.assertTrue(y == IntervalVector([[-1,1],[5,6]]))

if __name__ ==  '__main__':
  unittest.main()

File: tests/core/contractors/codac2_tests_CtcCartProd.cpp
/** 
 *  Codac tests
 * ----------------------------------------------------------------------------
 *  \date       2024
 *  \author     Simon Rohou
 *  \copyright  Copyright 2024 Codac Team
 *  \license    GNU Lesser General Public License (LGPL)
 */

#include <catch2/catch_test_macros.hpp>
#include <codac2_CtcCartProd.h>
#include <codac2_CtcInverse.h>

using namespace std;
using namespace codac2;

TEST_CASE("CtcCartProd")
{
  {
    VectorVar v(1);
    CtcInverse c1(AnalyticFunction({v}, v), IntervalVector({{-1,1}}));
    CtcInverse c2(AnalyticFunction({v}, v), IntervalVector({{5,6}}));

    IntervalVector a(1),b(1);
    c1.contract(a);
    CHECK(a == IntervalVector({{-1,1}}));
    c2.contract(b);
    CHECK(b == IntervalVector({{5,6}}));

    {
      IntervalVector x(2);
      CtcCartProd c3(c1,c2);
      c3.contract(x);
      CHECK(x == IntervalVector({{-1,1},{5,6}}));
    }

    {
      IntervalVector x(2);
      auto c4 = cart_prod(c1,c2);
      c4.contract(x);
      CHECK(x == IntervalVector({{-1,1},{5,6}}));
    }
  }
}